internal/controller/deploymenttrack/controller_finalize.go

```diff
@@ -0,0 +1,248 @@
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package deploymenttrack
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	choreov1 "github.com/choreo-idp/choreo/api/v1"
+	"github.com/choreo-idp/choreo/internal/controller"
+	"github.com/choreo-idp/choreo/internal/labels"
+)
+
+const (
+	// DeploymentTrackCleanupFinalizer is the finalizer that is used to clean up deployment track resources.
+	DeploymentTrackCleanupFinalizer = "core.choreo.dev/deploymenttrack-cleanup"
+)
+
+// ensureFinalizer ensures that the finalizer is added to the deployment track.
+// The first return value indicates whether the finalizer was added to the deployment track.
+func (r *Reconciler) ensureFinalizer(ctx context.Context, deploymentTrack *choreov1.DeploymentTrack) (bool, error) {
+	// If the deployment track is being deleted, no need to add the finalizer
+	if !deploymentTrack.DeletionTimestamp.IsZero() {
+		return false, nil
+	}
+
+	if controllerutil.AddFinalizer(deploymentTrack, DeploymentTrackCleanupFinalizer) {
+		return true, r.Update(ctx, deploymentTrack)
+	}
+
+	return false, nil
+}
+
+func (r *Reconciler) finalize(ctx context.Context, old, deploymentTrack *choreov1.DeploymentTrack) (ctrl.Result, error) {
+	finalizingCond := meta.FindStatusCondition(deploymentTrack.Status.Conditions, "Finalizing")
+
+	// First reconciliation - set the initial status
+	if finalizingCond == nil {
+		meta.SetStatusCondition(&deploymentTrack.Status.Conditions, NewDeploymentTrackStartFinalizeCondition(deploymentTrack.Generation))
+		if err := controller.UpdateStatusConditions(ctx, r.Client, old, deploymentTrack); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Execute appropriate action based on current status
+	switch finalizingCond.Reason {
+	case string(ReasonDeploymentTrackFinalizingStarted):
+		if err := r.deleteBuilds(ctx, deploymentTrack); err != nil {
+			return ctrl.Result{}, err
+		}
+		// Update status and requeue
+		meta.SetStatusCondition(&deploymentTrack.Status.Conditions, NewDeploymentTrackCleanBuildsCondition(deploymentTrack.Generation))
+		if err := controller.UpdateStatusConditionsWithPatch(ctx, r.Client, old, deploymentTrack); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
```

Comment:
Here, we don't need to requeue, the API server will anyway trigger another reconcile due to status update. Also, shall we use the Update instead of Patch? The patch will hide any conflicts that happen due to stale state. We can use the patch later for all other controllers if required.